src/colorizer/Dataset.ts

```diff
@@ -106,40 +88,77 @@ export default class Dataset {
 
     this.arrayLoader = arrayLoader || new JsonArrayLoader();
-    this.featureFiles = {};
-    this.features = {};
+    this.features = new Map();
     this.metadata = defaultMetadata;
   }
 
   private resolveUrl = (url: string): string => `${this.baseUrl}/${url}`;
 
-  private async fetchJson(url: string): Promise<DatasetManifest> {
+  private async fetchJson(url: string): Promise<AnyManifestFile> {
     const response = await urlUtils.fetchWithTimeout(url, urlUtils.DEFAULT_FETCH_TIMEOUT_MS);
     return await response.json();
   }
 
-  private async loadFeature(name: string, metadata: Partial<FeatureMetadata>): Promise<void> {
-    const url = this.resolveUrl(this.featureFiles[name]);
+  /**
+   * Parses a feature's `type` string and returns a FeatureType enum.
+   * @param inputType The `type` string to parse.
+   * @param defaultType Default value to return if `inputType` is not recognized.
+   * @returns The parsed FeatureType.
+   */
+  private getFeatureTypeFromString(inputType: string, defaultType: FeatureType = FeatureType.CONTINUOUS): FeatureType {
+    const type = inputType.toLowerCase();
+    switch (type) {
+      case "discrete":
+        return FeatureType.DISCRETE;
+      case "categorical":
+        return FeatureType.CATEGORICAL;
+      case "continuous":
+        return FeatureType.CONTINUOUS;
+      default:
+        return defaultType;
+    }
+  }
+
+  /**
+   * Loads a feature from the dataset, fetching its data from the provided url.
+   * @returns A promise of an array tuple containing the feature name and its FeatureData.
+   */
+  private async loadFeature(metadata: ManifestFile["features"][number]): Promise<[string, FeatureData]> {
+    const name = metadata.name;
+    const url = this.resolveUrl(metadata.data);
     const source = await this.arrayLoader.load(url);
-    this.features[name] = {
-      tex: source.getTexture(FeatureDataType.F32),
-      data: source.getBuffer(FeatureDataType.F32),
-      min: source.getMin(),
-      max: source.getMax(),
-      units: metadata?.units || "",
-    };
+    const featureType = this.getFeatureTypeFromString(metadata?.type || "", FeatureType.CONTINUOUS);
+    const featureCategories = metadata?.categories;
+    // Validation
+    if (featureType === FeatureType.CATEGORICAL && !metadata?.categories) {
+      throw new Error(`Feature ${name} is categorical but no categories were provided.`);
+    }
+    if (featureCategories && featureCategories.length > MAX_FEATURE_CATEGORIES) {
+      throw new Error(
+        `Feature ${name} has too many categories (${featureCategories.length} > max ${MAX_FEATURE_CATEGORIES}).`
+      );
+    }
+
+    return [
+      name,
+      {
+        tex: source.getTexture(FeatureDataType.F32),
+        data: source.getBuffer(FeatureDataType.F32),
+        min: source.getMin(),
+        max: source.getMax(),
+        units: metadata?.units || "",
+        type: featureType,
+        categories: featureCategories || null,
+      },
+    ];
   }
```

Comment:
Rewrote `loadFeature` so it returns the result instead of directly inserting into `this.features[name]`. This created a race condition that led to features being sorted differently each time they were loaded.